### ISSUE TITLE
FBXLoader doesn't support transformed pivot points

### ISSUE THREAD

**Issue by looeee** [mrdoob/three.js#11895]
Simplified from #11697 since I've figured out the issue.

Here is a simple animation i created in 3ds max:

![3dsmax](https://user-images.githubusercontent.com/5307958/29000095-5ae06a34-7a8b-11e7-99a1-cdb81ac824c0.jpg)

You can see the pivot point of the red bar has been moved to one end, and it is animated to swing around the pivot. 

Loading this in Clara.io:

![clara](https://user-images.githubusercontent.com/5307958/29000096-8431d616-7a8b-11e7-9701-e630ed9f2afc.jpg)

The animation plays correctly, and the pivot point is displayed in the correct place.

Loading in Paint3D in windows - animations are not played, but the bar is in the correct place:

![paint3d](https://user-images.githubusercontent.com/5307958/29000099-9fbbd5f8-7a8b-11e7-8fb7-64ba1e5420eb.jpg)

Finally, loading in three.js and I get this:

![three](https://user-images.githubusercontent.com/5307958/29000102-aa7131aa-7a8b-11e7-977b-a3a1d3d575e0.jpg)

The loader has assumed that the pivot point is at the centre of the bar and moved it accordingly - the animation plays around this new pivot. 

Here's the FBX file:

[simple joint.zip](https://github.com/mrdoob/three.js/files/1202827/simple.joint.zip)

**Comment by roccapl** [comment]
see #11878 maybe it's not a problem with the exporter, because I have a similar issue with the JSON format

**Comment by looeee** [comment]
@roccapl I investigated a bit and yes, transformed pivots are not supported by three.js. However Clara.io is built on three.js and the model loads fine there.

The obvious solution (to me) is that when loaders detect a transformed pivot, they create an empty Object3D to represent the pivot and make the mesh a child of that. Then all animations and transforms must be applied there as well. I'll look into this tomorrow to get an idea of how complicated it would be to implement.

**Comment by roccapl** [comment]
Could someone explain to me why this is unsupported and what do people do as a common practice to circumvent this limitation? I've seen plenty of complex skinned animations with complicated bone chains playing back with no problems in three, why is this case different?

**Comment by looeee** [comment]
I have found a solution! 🍾 🎆 
For this very simple jointed example anyway - it's not a fix for the FBXLoader so I'll keep it short here and any further discussion of this solution / workaround should take place on the forum.

But it seems the GLTF2Loader does support transformed pivots. So the process goes like this:

1. Install the [blender gltf2 exporter](https://github.com/KhronosGroup/glTF-Blender-Exporter/)
2. Export the file as FBX
3. Load the FBX file in blender
4. Export as GLTF2 ( used .glb binary format)
5. import in three.js and the results for this simple example look like this, and the animations play correctly (with a bit of tweaking of exporter settings)

![three-gltf](https://user-images.githubusercontent.com/5307958/29020421-2244874e-7b8c-11e7-9a17-6fde6259b0ae.jpg)

**Comment by looeee** [comment]
What this shows for this issue (and the related #11878 for JSON format issue), is that transformed pivots **can** be supported. So we need to find out how the GLTF2Loader (or exporter) supports them and see whether that will work of other loaders too.

**Comment by roccapl** [comment]
That's great news, I'll dig into the code on Thursday and see if I can come up with a patch. In the meantime if GLTF2 is a mature format I can try to use that instead of the JSON one.

**Comment by mrdoob** [comment]
Wouldn't know how to solve it. For your specific model this patch does the trick:

```patch
diff --git a/examples/js/loaders/FBXLoader.js b/examples/js/loaders/FBXLoader.js
index 9c2a64c..491c82b 100644
--- a/examples/js/loaders/FBXLoader.js
+++ b/examples/js/loaders/FBXLoader.js
@@ -1509,6 +1509,13 @@
 
 			}
 
+			if ( 'GeometricTranslation' in node.properties ) {
+
+				var array = node.properties.GeometricTranslation.value;
+				model.geometry.translate( array[ 0 ], array[ 1 ], array[ 2 ] );
+
+			}
+
 			var conns = connections.get( model.FBX_ID );
 			for ( var parentIndex = 0; parentIndex < conns.parents.length; parentIndex ++ ) {
 

```

<img width="1276" alt="screen shot 2017-08-08 at 9 01 45 am" src="https://user-images.githubusercontent.com/97088/29073074-41763022-7c18-11e7-81d6-ea41ce0daf4e.png">

However, this is modifying the geometry directly, so if other scene nodes use this same geometry things will break.

**Comment by looeee** [comment]
I created a simple rectangle in 3DS Max. By default the pivot is placed at the centre of the bottom face of the object. Then I centred the pivot to the object, and created two FBX files:

[Pivot test.zip](https://github.com/mrdoob/three.js/files/1210334/Pivot.test.zip)

Running these through a diff checker, aside from time stamps and IDs the only differences are 

**pivot_default.fbx** 

```
Objects:  { 
  ...
	Model: 2247454960, "Model::Box001", "Mesh" {
		Version: 232
		Properties70:  {
			P: "PreRotation", "Vector3D", "Vector", "",-90,-0,0
			P: "RotationActive", "bool", "", "",1
			P: "InheritType", "enum", "", "",1
			P: "ScalingMax", "Vector3D", "Vector", "",0,0,0
			P: "DefaultAttributeIndex", "int", "Integer", "",0
			P: "MaxHandle", "int", "Integer", "UH",1
		}
    ...
	}
  ...
}
```

**pivot_centered.fbx**
```
Objects:  { 
  ...
  Model: 2247454960, "Model::Box001", "Mesh" {
      Version: 232
      Properties70:  {
        P: "PreRotation", "Vector3D", "Vector", "",-90,-0,0
        P: "RotationActive", "bool", "", "",1
        P: "InheritType", "enum", "", "",1
        P: "ScalingMax", "Vector3D", "Vector", "",0,0,0
        P: "GeometricTranslation", "Vector3D", "Vector", "",0,0,-1.5
        P: "DefaultAttributeIndex", "int", "Integer", "",0
        P: "Lcl Translation", "Lcl Translation", "", "A",0,1.5,-0
        P: "MaxHandle", "int", "Integer", "UH",1
      }
    ...
	}
  ...
}
```

Specifically, centring the pivot adds the lines:

```
P: "GeometricTranslation", "Vector3D", "Vector", "",0,0,-1.5
P: "Lcl Translation", "Lcl Translation", "", "A",0,1.5,-0
```

Both of these files loaded in three.js display correctly, since it seems these lines are ignored and the pivot is assumed to be at the centre of the object. 

The problem arises when any animation is applied - animations will only be correct for **pivot_centered.fbx**, as all transformations will be applied around the centre of the object.

**Comment by looeee** [comment]
@mrdoob I can confirm that your solution works in my simple test case. For more complex examples though the model may be a Group or Object3D and this is required:

```js
if ( 'GeometricTranslation' in node.properties ) {

   var array = node.properties.GeometricTranslation.value;
    model.traverse( function( child ) {

        if( child.geometry ) {

           child.geometry.translate( array[ 0 ], array[ 1 ], array[ 2 ] );

       }

   } );
}
```

This looks better (unfortunately I can't share my model, but here is a before and after screenshot).

Without fix:
![pre_fix](https://user-images.githubusercontent.com/5307958/29105862-353eb5f2-7cfb-11e7-9945-86d9ba9bc03e.jpg)

With fix:
![post_fix](https://user-images.githubusercontent.com/5307958/29105865-39d21398-7cfb-11e7-8ef3-c12b0f3c3e0c.jpg)

The only part of this model that is currently animated are the legs which just swing back and forth, and as you can see some strange transform is applied to the left leg, while the right seems to be OK. 

Aside from that it looks like normals are inverted on the right leg and right arm.

So, still a lot of problems - but this gives me hope that most of them can be fixed in Max before exporting. 

>However, this is modifying the geometry directly, so if other scene nodes use this same geometry things will break.

At the moment, we are faced with a situation where nearly all complex animated or rigged models are broken. It might be the case that supporting these models is more useful than supporting models with reused geometry. But certainly more testing is in order before making that decision - and ideally a solution that suits both cases can be found.

**Comment by mrdoob** [comment]
Hmm... Would loading the file in Blender and exporting as gltf be an option?

**Comment by looeee** [comment]
Theoretically, yes - and it worked well for the person who made the linked forum post. However this is what happens when I try to load the FBX file in blender:

![blender](https://user-images.githubusercontent.com/5307958/29157595-bee8b29a-7dd1-11e7-92e3-26ccb2d26cfe.jpg)

and I have even less luck exporting it as collada instead of fbx. I'm in the process of trying other formats now though.

**Comment by looeee** [comment]
I found that exporting as .abc (alembic) format works , kind of. Materials are not exported, but the object does load in blender at least. However try as I might I cannot get my animations to export with the gltf2 exporter. Aside from this, gltf seems to have the same issue with winding order as #11911

![gltf](https://user-images.githubusercontent.com/5307958/29159949-9ae8cfe8-7dda-11e7-9993-5730724ca2c9.jpg)

**Comment by looeee** [comment]
Just wanna come back to this - it got a bit off topic, but as I understand it the issue at the moment is that we can either support animation using transformed pivot, or instanced geometry with non-centred pivots. And at the moment we are supporting instanced geometry. 

I think that supporting animations with transformed pivots is a much higher priority, given that its one of the most basic animation techniques. There's basically no way to get something as simple as a hinge joint working without this fix, let alone a complex rigged mesh. 

Instanced geometry with centred pivots will still work in any case, and if there is no way to deal with the instanced geometries with non-centred pivots, then that's a small price to pay for fixing animations.

**Comment by FishOrBear** [comment]
I thought I had the same problem

3dmax before exporting
![image](https://user-images.githubusercontent.com/19372111/29345739-3e16e15e-8273-11e7-8365-9a60e6d7637b.png)
After import


![image](https://user-images.githubusercontent.com/19372111/29345757-571b90aa-8273-11e7-9d95-056391399410.png)

There is also a problem that causes the imported grid to appear incorrectly.

[test.zip](https://github.com/mrdoob/three.js/files/1226960/test.zip)

[tv.zip](https://github.com/mrdoob/three.js/files/1226965/tv.zip)

**Comment by looeee** [comment]
@FishOrBear can you try implementing the [fix mentioned above](https://github.com/mrdoob/three.js/issues/11895#issuecomment-321154716) to see if this fixes your error?

**Comment by FishOrBear** [comment]
Thank you, I tried it.
Solved my question.
@looeee

**Comment by looeee** [comment]
Closed via #11963
